src/mono/mono/component/marshal-ilgen.c

```diff
@@ -0,0 +1,2859 @@
+
+		cb_to_mono->mb_emit_branch_label (mb, CEE_BR, label2);
+
+		cb_to_mono->mb_patch_branch (mb, label3);
+		cb_to_mono->mb_patch_branch (mb, label1);
+#endif
+		break;
+	}
+	default:
+		g_assert_not_reached ();
+	}
+	return conv_arg;
+}
+
+static gboolean
+emit_native_wrapper_validate_signature (MonoMethodBuilder *mb, MonoMethodSignature* sig, MonoMarshalSpec** mspecs)
+{
+	if (mspecs) {
+		for (int i = 0; i < sig->param_count; i ++) {
+			if (mspecs [i + 1] && mspecs [i + 1]->native == MONO_NATIVE_CUSTOM) {
+				if (!mspecs [i + 1]->data.custom_data.custom_name || *mspecs [i + 1]->data.custom_data.custom_name == '\0') {
+					cb_to_mono->mb_emit_exception_full (mb, "System", "TypeLoadException", g_strdup ("Missing ICustomMarshaler type"));
+					return FALSE;
+				}
+
+				switch (sig->params[i]->type) {
+				case MONO_TYPE_CLASS:
+				case MONO_TYPE_OBJECT:
+				case MONO_TYPE_STRING:
+				case MONO_TYPE_ARRAY:
+				case MONO_TYPE_SZARRAY:
+				case MONO_TYPE_VALUETYPE:
+					break;
+
+				default:
+					cb_to_mono->mb_emit_exception_full (mb, "System.Runtime.InteropServices", "MarshalDirectiveException", g_strdup_printf ("custom marshalling of type %x is currently not supported", sig->params[i]->type));
+					return FALSE;
+				}
+			}
+			else if (sig->params[i]->type == MONO_TYPE_VALUETYPE) {
+				MonoMarshalType *marshal_type = mono_marshal_load_type_info (mono_class_from_mono_type_internal (sig->params [i]));
+				for (guint32 field_idx = 0; field_idx < marshal_type->num_fields; ++field_idx) {
+					if (marshal_type->fields [field_idx].mspec && marshal_type->fields [field_idx].mspec->native == MONO_NATIVE_CUSTOM) {
+						cb_to_mono->mb_emit_exception_full (mb, "System", "TypeLoadException", g_strdup ("Value type includes custom marshaled fields"));
+						return FALSE;
+					}
+				}
+			}
+		}
+	}
+
+	return TRUE;
+}
+
+static int
+emit_marshal_ptr_ilgen (EmitMarshalContext *m, int argnum, MonoType *t,
+		  MonoMarshalSpec *spec, int conv_arg,
+		  MonoType **conv_arg_type, MarshalAction action)
+{
+	MonoMethodBuilder *mb = m->mb;
+	switch (action) {
+	case MARSHAL_ACTION_CONV_IN:
+		/* MS seems to allow this in some cases, ie. bxc #158 */
+		/*
+		if (MONO_TYPE_ISSTRUCT (t->data.type) && !mono_class_from_mono_type_internal (t->data.type)->blittable) {
+			char *msg = g_strdup_printf ("Can not marshal 'parameter #%d': Pointers can not reference marshaled structures. Use byref instead.", argnum + 1);
+			cb_to_mono->mb_emit_exception_marshal_directive (m->mb, msg);
+		}
+		*/
+		break;
+
+	case MARSHAL_ACTION_PUSH:
+		cb_to_mono->mb_emit_ldarg (mb, argnum);
+		break;
+
+	case MARSHAL_ACTION_CONV_RESULT:
+		/* no conversions necessary */
+		cb_to_mono->mb_emit_stloc (mb, 3);
+		break;
+
+	default:
+		break;
+	}
+	return conv_arg;
+}
+
+static int
+emit_marshal_boolean_ilgen (EmitMarshalContext *m, int argnum, MonoType *t,
+		      MonoMarshalSpec *spec,
+		      int conv_arg, MonoType **conv_arg_type,
+		      MarshalAction action)
```

Comment:
tab/spaces mix

Reply:
There are are more. I would suggest fixing that for all the changes here.

Reply:
I didn't actually make marshal-ilgen in this change, its just been moved. But I guess it makes sense to fix these while I am at it.